tools/clang/lib/Sema/SemaHLSLDiagnoseTU.cpp

```diff
@@ -295,6 +301,39 @@ clang::FunctionDecl *ValidateNoRecursion(CallGraphWithRecurseGuard &callGraph,
   return nullptr;
 }
 
+bool DiagnoseHLSLMethodCall(CXXMemberCallExpr *CE, clang::Sema *self) {
+  const CXXMethodDecl *MD = CE->getMethodDecl();
+  if (MD->hasAttr<clang::HLSLIntrinsicAttr>()) {
+    hlsl::IntrinsicOp opCode =
+        (IntrinsicOp)MD->getAttr<clang::HLSLIntrinsicAttr>()->getOpcode();
+    if (opCode == hlsl::IntrinsicOp::MOP_CalculateLevelOfDetail ||
+        opCode == hlsl::IntrinsicOp::MOP_CalculateLevelOfDetailUnclamped) {
+      if (isIllegalIntrinsic(MD, opCode, self)) {
```

Comment:
The splitting of logic here is a bit odd. `isIllegalIntrinsic` has the logic for why the intrinsic is not allowed, but the reporting of the error is here disjointed from the logic. That makes this code significantly harder to follow, reuse and extend.

Reply:
Fixed.